### ISSUE TITLE
New game fails to start after upgrading FluidSynth to version 2.2.0

### ISSUE THREAD

**Issue by ghisvail** [OpenTTD/OpenTTD#9117]
OpenTTD works fine with FluidSynth version 2.1.8. When I upgrade its Flatpak runtime with FluidSynth version 2.2.0, the game is no longer playable.

## Version of OpenTTD

OpenTTD version 1.11.1 installed with Flatpak on Debian 11.

## Expected result

New game starts successfully.

## Actual result

New game starts but the progress bar of the initialization process hangs towards the end and the user interface becomes frozen. No way to escape the game apart from killing the app.

## Steps to reproduce

Install the game with FluidSynth version 2.2.0, launch OpenTTD, start a new game.

**Comment by glx22** [comment]
It seems there are some breaking changes in fluidsynth 2.2.0 https://www.fluidsynth.org/api/RecentChanges.html#NewIn2_2_0

**Comment by James103** [comment]
Expecially note the following two lines of the changelog:
* fluid_file_callbacks_t now uses **long long** as file-offset type (see fluid_long_long_t). This is a breaking change, which allows to load SoundFonts bigger than 2GiB on Windows. This change required to **bump fluidsynth's SOVERSION**.
* Various fluid_event_*() functions that received a "value" argument of type **short** now receive an **int** argument in preparation for MIDI 2.0 support

**Comment by ghisvail** [comment]
Although these are breaking changes from an ABI standpoint, the app still compiles and launches fine. I can see the start screen, navigate the menu and listen to the intro song. It's only the new game process which ultimately hangs for some reason.

**Comment by glx22** [comment]
I tried running it in gdb, and I can see it hangs on `fluid_player_join()` in `MusicDriver_FluidSynth::StopSong()`, but I don't know why.

**Comment by ghisvail** [comment]
On a side note, according to the [docs](https://www.fluidsynth.org/api/group__midi__player.html#ga6a953d057b72855d2e366d324a5d04e0), `fluid_player_join` always return `FLUID_OK` so the following [line](https://github.com/OpenTTD/OpenTTD/blob/e7cf589868480022599f6935997ef45963fc3f5a/src/music/fluidsynth.cpp#L172) is dead code.

**Comment by glx22** [comment]
Oh BTW I only tested with version available in `flatpak install openttd` so I don't know which version of fluidsynth is used, but it hangs.

**Comment by ghisvail** [comment]
There is an update coming which downgrades FluidSynth to version 2.1.8 in the runtime. If it hangs, then you are still using the runtime with version 2.2.0 which was pushed last night.

**Comment by LordAro** [comment]
Even Arch is not yet updated to fluidsynth 2.2 ! :)

**Comment by ghisvail** [comment]
> Even Arch is not yet updated to fluidsynth 2.2 ! :)

Well then, they should not :-)

**Comment by ghisvail** [comment]
@glx22 If needed, I can provide a build with FluidSynth v2.2.0 in a separate PR for testing. Please let me know.

The Flatpak version is back to version 2.1.8 for now.

**Comment by rubidium42** [comment]
Building both Fluidsynth 2.2.0 and OpenTTD from source on Debian Bullseye results, for me, in a working OpenTTD with music (original Windows). 

I have no idea what Flatpak is doing, but if it has built OpenTTD against the 2.1.8 version of Fluidsynth and then links/distributes the 2.2.0 version of Fluidsynth that is asking for problems due to the incompatible API/ABI change. Practically there is not that much OpenTTD can do about that, except during build time get the version of the library and checking whether the run time library seems to be compatible (if that information is even available during run time). Having said that, ICU is notorious for changing their API in incompatible ways but they ensure themselves that you cannot build and run with different ABIs.

Or maybe Fluidsynth fails itself on the music set that you have chosen?

**Comment by glx22** [comment]
I tried building manually Fluidsynth from source (git clone), and building OpenTTD against 2.1.8 works fine, while against 2.2.0 it hangs at the end of map generation, waiting for music to stop. Maybe related to https://github.com/FluidSynth/fluidsynth/commit/8413c35aca641567baf13e9b16e9839019ebf99d

**Comment by rubidium42** [comment]
Ah, figured out what I was doing wrong. The problem is introduced by https://github.com/FluidSynth/fluidsynth/commit/676923757c4f96c5c34931da5566d5b5be3e15eb. The issue that glx22 mentioned is a fix to solve something the commit I linked broke, but not our use case. As far as I could figure out the new "stopping" state never progresses to "stopped" for us, so the join that waits until it is "stopped" will never return.

I made an upstream bug report: https://github.com/FluidSynth/fluidsynth/issues/872. Maybe our code is wrong, in which case we still have something to fix, but I hope the developers of FluidSynth will tell us what is wrong about that code.
